### DISCUSSION TITLE
Any plans to upgrade to go 1.18?

### DISCUSSION THREAD

**Discussion by vedadiyan** [nats-io/nats.go#1139] [Ideas]
Is there any plan to upgrade to go 1.18 so that generics can be introduced for Encoded Connections? That allows for a generic Request function that would be more convenient to use.  

I would like create a pull request for a version that I modified and I added a few functionalities for Encoded Connections which require go 1.18  

1- A generic version of `Request` which returns a wrapped result under `EncodedMsg`
2- I added `Headers` to the new `EncodedMsg` struct 
3- I added another version of `Publish` to accept `EncodedMsg` so that headers could be used as well 

I required these features, because I would like to somehow deliver a response when something goes wrong in the subscriber and a response cannot be produced. Then I can report the error through headers. But the existing version does not allow that. (Headers are available for plain connections and not encoded connections)

**Comment by derekcollison** [comment]
We are exploring a bit, we would not use the encoded conns and more likely would introduce a better way of chaining handlers together, similar to "middleware" for HTTP.
